### PR TITLE
Support closed-over data-dependent indexing in scan (forward and autograd)

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186139
* #186138

A scan combine_fn that closes over a tensor and indexes it with the per-step
integer xs element (e.g. table[:, t], a common JAX idiom in HMMs and decoding
loops) previously failed, while jax.lax.scan handles it. This builds on
data-dependent 0-dim tensor indexing (parent commit) and adds the
scan-specific pieces for both forward and autograd.

Forward: scan traces the body with a fake per-step slice whose value is an
unbacked SymInt. Such body-internal symbols never surface in scan's outputs
(whose shapes are consistent across iterations), so the fake-tensor-mode body
run now ignores fresh unbacked symbols (ignore_fresh_unbacked_symbols).

Backward (review these two together):
  - The functionalized select_copy.int had no data-dependent-friendly meta and
    fell through to the hard-guarding C++ impl. meta_select_copy mirrors
    meta_select's bounds but returns a contiguous copy, so it has no
    storage-offset symbol to reason about.
  - The partitioner saved the per-step index value (an unbacked SymInt from
    .item()) as a forward intermediate, which scan stacks across iterations --
    impossible for a SymInt. Unbacked basic symbols are now recomputed in
    backward (they are recomputable from the read-only inputs), alongside the
    existing complex-expr recompute.

Test Plan:
```
python test/functorch/test_control_flow.py TestControlFlow -k test_scan_closed_over_indexing
python test/functorch/test_control_flow.py TestControlFlow -k scan
python test/functorch/test_control_flow.py TestControlFlowTraced -k scan
```

This PR was authored with the assistance of an AI coding assistant (Claude).

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review these changes
